CNI: move CNI configuration management to the agent, convert to Hive

squeed opened the pull request:
This is the new version of #22035, which staled out. The functionality is the same.

This PR moves CNI configuration file management to the agent and eliminates the `cni-install.sh` postStart hook. By and large, the functionality is identical with one exception around timing (we now watch instead of retry).

For more detailed information, see the individual commits. A short summary of the changes:
* the CNI configuration manager is now a Cell in the Hive
* Instead of retry loops, we now watch the CNI configuration directory for changes
* In the event of another CNI configuration file being written, we will immediately remove it, rather than on agent restart
* Customization by environment variables is removed (we never _actually_ supported this anyways)

```release-note
The Cilium agent now manages the CNI configuration file. This will allow for faster startup times when injecting Cilium as a chained plugin, such as with aws-cni.
```

Fixes: #14892
Fixes: #23730

squeed committed:
daemon / cni: move to Cell, watch for changes

Refactor the CNI controller to use the Hive / Cell mechanism. This means
simplifying the underlying controller and making it run immediately,
rather than waiting for the Daemon to start. This mirrors the existing
behavior of the postStart hook, which will be going away.

The controller now also watches the CNI configuration directory and
reconciles whenever a file has changed; this is needed because some
chaining modes (e.g. AWS) need to wait until the aws-cni plugin installs
its own CNI configuration file. We then want to install our chained
plugin ASAP.

Previously, we would just loop until the configuration was created, but
that is fragile. Rather, set up an fsnotify and watch for changes that
way. This has the advantage of re-generating the chained configuration should
the underlying network change.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
helm: remove cni-install.sh / enable agent CNI configuration management

Move all CNI configuration management exclusively to the agent.
Previously, it was split between the agent and a preStart hook. Now that
the agent implements all the functionality, keep it there.

This removes the cni-install.sh preStart hook, clarifies some
documentation, and sets the write-cni-conf-when-ready flag in (almost)
all cases, enabling CNI management in the agent.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

qmonnet commented on Documentation/network/kubernetes/configuration.rst:
```diff
@@ -150,65 +144,59 @@ The CNI auto installation is performed as follows:
-    wget https://storage.googleapis.com/kubernetes-release/network-plugins/cni-0799f5732f2a11b329d9e3d51b9c8f2e3759f2ff.tar.gz
-    sudo tar -xvf cni-0799f5732f2a11b329d9e3d51b9c8f2e3759f2ff.tar.gz -C /opt/cni
-    rm cni-0799f5732f2a11b329d9e3d51b9c8f2e3759f2ff.tar.gz
+3. The file ``/etc/cni/net.d/05-cilium.conflist`` is written.
```
Does this mean that `/etc/cni/net.d/05-cilium.conf` no longer exists? Asking because I can still observe several occurrences of `05-cilium.conf` (not `...conflist`) in the repo.

Is it worth a note in `Documentation/operations/upgrade.rst`?

squeed replied:
Very good observation on both points!

squeed replied:
That is to say, you're correct, `05-cilium.conf` is no longer being written. I'll also update the PR to remove any "old" or stale configuration files.

qmonnet replied:
Thanks!

I can still observe the following:

```
$ rg -w '05-cilium.conf'
[...]
Documentation/installation/cni-chaining-portmap.rst
53:``05-cilium.conflist`` and remove the standard ``05-cilium.conf``. The new

install/kubernetes/cilium/files/nodeinit/startup.bash
101:    echo -e "{\n\t"cniVersion": "0.3.1",\n\t"name": "cilium",\n\t"type": "cilium-cni"\n}" > /etc/cni/net.d/05-cilium.conf
```

Please update if relevant.

Not sure if you agreed on adding a note to `Documentation/operations/upgrade.rst` (I believe it would be a good thing), but I can't find one in the updated PR.

squeed replied:
Whoops, forgot to `git add` :-p

squeed replied:
The entry in `startup.bash` is a placeholder; there's no need to change it.

qmonnet replied:
Ok, thank you for the updates!

qmonnet commented on Documentation/network/kubernetes/configuration.rst:
```diff
@@ -150,65 +144,59 @@ The CNI auto installation is performed as follows:
 
-The CNI configuration file is automatically written and maintained by the
-scripts ``cni-install.sh`` and ``cni-uninstall.sh`` which are running as
-``postStart`` and ``preStop`` hooks of the Cilium pod.
```
Does removing this has any impact on https://github.com/cilium/cilium/pull/24375?

squeed replied:
Oh, good question! I took a look and: no, we're good. That PR allows for tweaking the nodeinit daemonset, not the preStart hook for the agent itself.

qmonnet commented on Documentation/network/kubernetes/configuration.rst:
```diff
@@ -150,65 +144,59 @@ The CNI auto installation is performed as follows:
++----------------+----------------------------------------+---------+
+| Helm variable  | Description                            | Default |
++================+========================================+=========+
+| cni.customConf | Disable CNI configuration management   | false   |
```
Nit: Let's format this as code? Here and for the other entries

```suggestion
| ``cni.customConf`` | Disable CNI configuration management   | false   |
```

squeed replied:
done

qmonnet reviewed (commented):
Looks good, even though I looked mostly at the docs. I've got a few questions below:

squeed committed:
cmd/cleanup: also remove 05-cilium.conflist

We can also write this CNI configuration file, so add it to the list of
files to clean up.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

qmonnet commented on Documentation/network/kubernetes/configuration.rst:
```diff
@@ -150,65 +144,59 @@ The CNI auto installation is performed as follows:
++======================+================================================================+
+| ``cni.readCniConf``  | Path (inside the agent) to a source CNI configuration file     |
++----------------------+---------------------------------------------------------------+
+| ``cni.configMap``    | Name of a ConfigMap containing a source CNI configuration file |
```
I think the missing dash here is what the doc/Netlify workflows complain about:

```suggestion
| ``cni.readCniConf``  | Path (inside the agent) to a source CNI configuration file     |
+----------------------+----------------------------------------------------------------+
| ``cni.configMap``    | Name of a ConfigMap containing a source CNI configuration file |
```

qmonnet reviewed (approved):
Looks good to me, provided CI issues get solved. Thanks!

sayboras reviewed (approved):
The CNI code is pretty much easier to follow now :heavy_check_mark:

nathanjsweet commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -62,7 +62,9 @@ cilium-agent [flags]
       --clustermesh-config string                               Path to the ClusterMesh configuration directory
-      --cni-chaining-mode string                                Enable CNI chaining with the specified plugin
+      --cni-chaining-mode string                                Enable CNI chaining with the specified plugin (default "none")
+      --cni-exclusive                                           Whether to remove other CNI configurations
```
What are your thoughts on documenting [the default value](https://github.com/cilium/cilium/blob/c3bed624dee6f9a6aa7c5abaa933f8b4745e6467/install/kubernetes/cilium/values.yaml.tmpl#L486) for this option?

squeed replied:
Hmm, not sure what you mean. The helm value hasn't changed; previously it was consumed by a sketchy and undocumented shell script; now the agent is responsible.

nathanjsweet replied:
Yes, but it was only a helm value before. Now it's a Cilium command line value.

squeed replied:
huh, all that is auto-generated. I wonder why it doesn't automatically document a default value (I supply one).

nathanjsweet commented on Documentation/network/kubernetes/configuration.rst:
```diff
@@ -129,17 +129,11 @@ CNI
-          instructions on how to install :term:`CNI` in case the ``loopback`` binary
-          is not already installed on your worker nodes.
-
 CNI configuration is automatically being taken care of when deploying Cilium
```
I know this isn't part of your commit, but is there any chance you can fix it?

```suggestion
CNI configuration is automatically taken care of when deploying Cilium
```

nathanjsweet commented on Documentation/network/kubernetes/configuration.rst:
```diff
@@ -150,65 +144,59 @@ The CNI auto installation is performed as follows:
 Adjusting CNI configuration
 ---------------------------
 
-The CNI configuration file is automatically written and maintained by the
```
Did you mean to remove this line?

squeed replied:
yeah; this PR gets rid of the postStart hook

squeed replied:
ohh nevermind you're right.

nathanjsweet commented on Documentation/network/kubernetes/configuration.rst:
```diff
@@ -150,65 +144,59 @@ The CNI auto installation is performed as follows:
-environment variables can be specified in the :term:`DaemonSet` file like this:
-
-.. code-block:: yaml
+cilium pod. It is written after the agent has finished initialization and
```
"cilium pod" is a fragment. I'm assuming you meant to keep the first deleted line above?

nathanjsweet commented on Documentation/network/kubernetes/configuration.rst:
```diff
@@ -150,65 +144,59 @@ The CNI auto installation is performed as follows:
+| ``cni.configMapKey`` | Install CNI configuration and binaries                         |
++----------------------+----------------------------------------------------------------+
+
+These Helm variables are converted to a smaller set of cilium ConfigMap keys:
```
`cni-exclusive` is a settable helm value. Do you mean to make its value programmatically set?

squeed replied:
yeah.

Honestly, this whole document is just really old. We don't document any other helm variables in the same manner, but I was trying to preserve the existing style for CNI configuration.

nathanjsweet reviewed (changes requested):
Looking good. I have a few nits and one point of confusion on my part.

nathanjsweet reviewed (changes requested):
Nit if you want it.
